### PR TITLE
Use custom types for immediate update

### PR DESCRIPTION
**Pull request by xsebek**
- make `WorldUpdate` and `RobotUpdate` serializable
- part of #50

### CODE REVIEW

File: src/Swarm/Game/CESK.hs [hunk 1]
```diff
@@ -350,39 +352,22 @@ prettyFrame (FDef x) = "def " ++ from x ++ " = _"
...
-  }
-
-newtype RobotUpdate = RobotUpdate
-  { robotUpdateInventory :: Inventory -> Inventory
+data WorldUpdate = ReplaceEntity
+  { updatedLoc :: V2 Int64
+  , originalEntity :: Entity
+  , newEntity :: Maybe Entity
   }
+  deriving (Eq, Ord, Show, Generic, FromJSON, ToJSON)
```
Comment (xsebek) [inline src/Swarm/Game/CESK.hs:368]: The default instances do not follow our style, so I would like to customize it a bit and test it before merging.

File: src/Swarm/Game/CESK.hs [hunk 2]
```diff
@@ -350,39 +352,22 @@ prettyFrame (FDef x) = "def " ++ from x ++ " = _"
 prettyFrame FExec = "exec _"
 prettyFrame (FBind Nothing t _) = "_ ; " ++ prettyString t
 prettyFrame (FBind (Just x) t _) = from x ++ " <- _ ; " ++ prettyString t
-prettyFrame FImmediate {} = "(_ : cmd a)"
+prettyFrame (FImmediate c w r) = unwords ["immediate@" <> T.unpack (Syntax.syntax $ Syntax.constInfo c), show w, show r]
 prettyFrame (FUpdate loc) = "store@" ++ show loc ++ "(_)"
 prettyFrame FFinishAtomic = "finishAtomic"
 
 --------------------------------------------------------------
 -- Wrappers for functions in FImmediate
```
Comment (byorgey) [inline src/Swarm/Game/CESK.hs:360]: We should get rid of this comment too, `FImmediate` no longer contains functions.

File: src/Swarm/Game/CESK.hs [hunk 3]
```diff
@@ -164,7 +166,7 @@ data Frame
     --   if there is one, to the output of the first command).
     FBind (Maybe Var) Term Env
   | -- | Apply specific updates to the world and current robot.
-    FImmediate WorldUpdate RobotUpdate
+    FImmediate Const [WorldUpdate] [RobotUpdate]
```
Comment (byorgey) [inline src/Swarm/Game/CESK.hs:169]: Is the `Const` just for better error messages?  So we know what command was ultimately responsible for generating this `FImmediate` frame?

**Review by byorgey — APPROVED** [review]
Looks good.  Not only does this make it possible to serialize `WorldUpdate` and `RobotUpdate`, I think the code is also much nicer now!

File: src/Swarm/Game/CESK.hs (hunk 3 above)
Comment (xsebek) [reply src/Swarm/Game/CESK.hs:169]: Exactly, `FImmediate` was definitely created by some command which we use in the error message.

I will note this in the docstring. 👍

File: src/Swarm/Game/CESK.hs (hunk 2 above)
Comment (xsebek) [reply src/Swarm/Game/CESK.hs:360]: Indeed, I am considering moving them to Robot and World respectively anyway. 🙂

File: src/Swarm/Game/CESK.hs (hunk 1 above)
Comment (xsebek) [reply src/Swarm/Game/CESK.hs]: Scratch that, doing it here would blow up the scope and I want to clean up first.

**Comment by xsebek** [comment]
Happy New Year to Swarm! :partying_face: @byorgey @kostmo

**Pull request merged**